Adding kustomize templates for k8s onboarding (#4754)
elastic/elastic-agent: deploy/kubernetes/elastic-agent-kustomize/default/README.md

- %CA_TRUSTED%: The ssl.ca_trusted_fingerprint in order the elastic agent to be able to trust the certificate authority of the Elasticsearch output.
- %DATASET_ID%: A string that will be added as a new field and will denote a specific installation. *By default, this will be added to state_pod dataset.*

## Remote usage of kustomize templates

Users can use following commands:

Managed Elastic Agent:

```bash
❯ kubectl https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-maanged\?ref\=main | sed -e "s/JUVOUk9MTE1FTlRfVE9LRU4l/base64_ENCODED_ENROLLMENT_TOKEN/g" -e "s/%FLEET_URL%/https:\/\/localhost:9200/g" | kubectl apply -f-

```

Standalone Elastic Agent:

```bash
kubectl kustomize https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-standalone\?ref\=main | sed -e "s/JUFQSV9LRVkl/<base64_encoded_APIKEY>/g" -e "s/%ES_HOST%/https:\/\/localhost:9200/g" -e "s/%CA_TRUSTED%/ca_trusted_fingerprint/g" -e "s/%DATASET_ID%/12345/g" | kubectl apply -f-
```

Examples of Base64 encoded values:

```bash
❯ echo %API_KEY | base64
JUFQSV9LRVkK

> echo %ENROLLMENT_TOKEN% | base64

Reviewer: here should be used `echo -n` instead, otherwise it might not work:
```
% echo -n "Zlh5ekw1QUJubU8xdUFJajNpS2M6QWdaZFIwX0xUNmFhQ3BGbUVxTVZwUQ==" | base64
WmxoNWVrdzFRVUp1YlU4eGRVRkphak5wUzJNNlFXZGFaRkl3WDB4VU5tRmhRM0JHYlVWeFRWWndVUT09

% echo "Zlh5ekw1QUJubU8xdUFJajNpS2M6QWdaZFIwX0xUNmFhQ3BGbUVxTVZwUQ==" | base64
WmxoNWVrdzFRVUp1YlU4eGRVRkphak5wUzJNNlFXZGFaRkl3WDB4VU5tRmhRM0JHYlVWeFRWWndVUT09Cg==
```

when the result of `echo <token> | base64` is used, agent fails with error:
```
Error: fail to enroll: fail to execute request to fleet-server: net/http: invalid header field value for "Authorization"
```

ideally would be to avoid this manual transformation, wondering if kustomize does not provide this option somehow

Author: @tetianakravchenko  I update the README file for this. Indeed echo adds an extra line by default. 

During the onboarding flow the base64 values are provided from the kibana ui page

# Kustomize Templates

The list below includes the official [kustomize](https://github.com/kubernetes-sigs/kustomize) templates to run them in Kubernetes:

Agent Scenario | Description
---- | ----
[Elastic Agent managed - Default ](./elastic-agent-kustomize/default/elastic-agent-managed/) | Default Elastic Agent managed by Fleet setup. Kube-state-metrics (KSM) is installed automatically.

> [Elastic Agent standalone Default ](./elastic-agent-kustomize/default/elastic-agent-standalone/) | Default Standalone Elastic Agent setup. Kube-state-metrics (KSM) is installed automatically.

Reviewer 2: Better to link from `https://github.com/elastic/elastic-agent/...`, otherwise 404

## Using above templates

Users can clone this repository to use the provided kustomize templates.

> For *Managed Elastic Agent*, please update the following secrets inside main kustomization.yaml:

Reviewer 2: Here it is better to link the file.

Author: Done

- %FLEET_URL%: Fleet Server URL to enroll the Elastic Agent into. FLEET_URL can be found in Kibana, go to Management > Fleet > Settings
- %ENROLLMENT_TOKEN%: Elasticsearch API key used to [enroll Elastic Agents](https://www.elastic.co/guide/en/fleet/current/fleet-enrollment-tokens.html#fleet-enrollment-tokens) in Fleet. *This should be encoded as base64 value because it will be stored as Kubernetes secret*

Eg.

```yaml
secretGenerator:
  - name: elastic-agent-creds
    literals:
      - enrollment_token=%ENROLLMENT_TOKEN%
```

> Comm
> 

Reviewer 3: ```suggestion
```

This doesn't look like it is supposed to be here.

For *Standalone Elastic Agent*, please update the following secrets inside main [kustomization.yaml](./elastic-agent-managed/kustomization.yaml):

- %ES_HOST%: The Elasticsearch host to communicate with

> - %API_KEY: The API Key with access privilleges to connect to Elasticsearch. See [create-api-key-standalone-agent](https://www.elastic.co/guide/en/fleet/current/grant-access-to-elasticsearch.html#create-api-key-standalone-agent). *This should be encoded as base64 value because it will be stored as Kubernetes secret*

Reviewer 3: ```suggestion
- %API_KEY: The API Key with access privileges to connect to Elasticsearch. See [create-api-key-standalone-agent](https://www.elastic.co/guide/en/fleet/current/grant-access-to-elasticsearch.html#create-api-key-standalone-agent). *This should be encoded as base64 value because it will be stored as Kubernetes secret*
```

> - %CA_TRUSTED%: The ssl.ca_trusted_fingerprint in order the elastic agent to be able to trust specific certificate

Reviewer 4: Is this needed to support self-signed certificates? We can skip it for production use-cases, right?

Author: Indeed it is needed only for self-signed and can be skipped for production use-cases.
We can run a test with a "production" site and prove that if we leave it that wont create any problem

> - %CA_TRUSTED%: The ssl.ca_trusted_fingerprint in order the elastic agent to be able to trust specific certificate

Reviewer 3: ```suggestion
- %CA_TRUSTED%: The ssl.ca_trusted_fingerprint in order the elastic agent to be able to trust the certificate authority of the Elasticsearch output.
```

> - %ELASTIC_AGENT_ID%: A numeric ELASTIC_AGENT_ID that will be added as a new field. *By default, this will be added to state_pod dataset.*

Reviewer 4: Not sure if this is the right name for this variable, real Elastic Agent ID is not known at the time we run `kubectl ...`, something like `DATASET_ID` might work better.

Author: To be accurate this is the session id that will be created from kibana and will denote the specific agent or in general the installation process. Yes I dont mind to change it to whatever UI team things is more appopriate

Reviewer 4: Awesome, thank you! `SESSION_ID` sounds to me a bit ephemeral, considering that it will be present for the whole lifespan of the agent config. `DATASET_ID` is the best I can think of now…

Oh, and btw, could we make it a string type so it supports UUID values?

Author: Thanks done: [111d8d0](https://github.com/elastic/elastic-agent/pull/4754/commits/111d8d0a3cb47e2941faf049e8eb3c4f20051bbf)

> - %DATASET_ID%: A string that will be added as a new field and will denote a specific installation. *By default, this will be added to state_pod dataset.*

Reviewer 3: I don't understand what this does. Is this dataset in the context of the data stream naming scheme? Why is there only one?

Author: In the standalone agent installations we can not verify if agents is installed and when is ready. So when a user through kibana will start a new installation (so it will have a specific session), we will include this session in the field dataset_id. And then kibana will have a loop to check whether this field has been created. 

FYI see https://github.com/elastic/elastic-agent/issues/4657 - the [Stretched goal]

Reviewer 3: I name that doesn't repurpose the dataset terminology would make this clearer.

Also this is effectively an agent ID that isn't set by Fleet? We could just add the ability to set that so that it would populate as the `agent.id` field in documents instead of us generating a random one at startup.

Author: > name that doesn't repurpose the dataset terminology would make this clearer.

@flash1293 , @mykolaharmash any other ideas regarding variable name? 

@cmacknz actually yes this is an agentId for standalone scenarios. 

> We could just add the ability to set that so that it would populate as the agent.id

When this is ready, I agree we can remove this workaround. This is I guess a quick solution now to make this working. Do you have some implementation in mind?

Reviewer 5: It's not really an agent id, because it will be the same on every node of the cluster - by my understanding, the individual nodes will get assigned separate agent ids.

I agree that "dataset" is a bit confusing too, what about "onboarding id" - that's what it is after all

Reviewer 3: Anything that isn't already a term used elsewhere in the product is fine with me.

Author: Thank you both, ONBOARDING_ID added in 
[89cc0d9](https://github.com/elastic/elastic-agent/pull/4754/commits/89cc0d9062d59634111d18f6a83353f75783ef56)

Managed Elastic Agent:

```bash
kubectl kustomize https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-managed\?ref\=main | 
sed -e 's/JHtlbnJvbGxtZW50X3Rva2VufQo=/<base64_fleet_enrollment_token>/g' -e 's/JHtmbGVldF91cmx9Cg==/<base64_fleet_url>/g'  | 
kubectl apply -f-
```

Standalone Elastic Agent:

```bash
kubectl kustomize https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-standalone\?ref\=main | 
sed -e 's/JHthcGlfa2V5fQo=/<base64_api_key>/g' -e 's/JHtlc19ob3N0fQo=/<base64_es_host>/g'  | 
kubectl apply -f-
```

> > NOTE: Base64 values

Reviewer 2: The format is not correct here, maybe it needs an empty line before the bash code?

Author: fixed thanks!

> kubectl kustomize https://github.com/elastic/elastic-agent/deploy/kubernetes/elastic-agent-kustomize/default/elastic-agent-standalone\?ref\=main | sed -e "s/JUFQSV9LRVkl/<base64_encoded_APIKEY>/g" -e "s/%ES_HOST%/https:\/\/localhost:9200/g" -e "s/%CA_TRUSTED%/ca_trusted_fingerprint/g" -e "s/%DATASET_ID%/12345/g" | kubectl apply -f-

Reviewer 4: Just to double check, should `sed -e "s/JUFQSV9LRVkl/<base64_encoded_APIKEY>/g"` be sed -e "s/%API_KEY%/<base64_encoded_APIKEY>/g"?

Reviewer 4: Ah, I see, `JUFQSV9LRVkl` is indeed inside the config produced by `kustomize`. @gizas could you please clarify what is this value, can we rely on it not changing for the purpose of onboarding?

Author: Indeed you dont need to change it. 
Actually initially the kustomize will create a secrtet api_key with encoded value of JUFQSV9LRVk
The reason for this is that the creation of variables as secrtes is done in base64 
```yaml
apiVersion: v1
data:
  api_key: JUFQSV9LRVkl
kind: Secret
```

Reviewer 4: Ah, I see, it's just `%API_KEY%` encoded with base64, thank you!

    ```bash
    ❯ echo '${api_key}' | base64
    JHthcGlfa2V5fQo=
    ❯ echo '${fleet_url}' | base64
    JHtmbGVldF91cmx9Cg==
    >echo '${enrollment_token}' | base64
    JHtlbnJvbGxtZW50X3Rva2VufQo=
    ❯ echo '${es_host}' | base64
    JHtlc19ob3N0fQo=
    ❯ echo JHtlc19ob3N0fQo= | base64 -D
    ${es_host}
    ```

## Updating kustomize templates

The included kustomize templates are being produced based on [Makefile](deploy/kubernetes/Makefile) by relevant automation: `GENERATEKUSTOMIZE=true make ci-create-kustomize`

> The current templates are using patches as defined [here](./elastic-agent-kustomize/default/elastic-agent-managed/kustomization.yaml)

Reviewer 2: Also needs another link (because this PR is not merged, maybe you need to write it as if exists in main already)

# Kustomize Templates

The list below includes the official [kustomize](https://github.com/kubernetes-sigs/kustomize) templates to run them in Kubernetes:

Agent Scenario | Description
---- | ----
[Elastic Agent managed - Default ](./elastic-agent-managed/) | Default Elastic Agent managed by Fleet setup. Kube-state-metrics (KSM) is installed automatically.
[Elastic Agent standalone Default ](./elastic-agent-standalone/) | Default Standalone Elastic Agent setup. Kube-state-metrics (KSM) is installed automatically.

## Using above templates

Users can clone this repository to use the provided kustomize templates.

> For *Managed Elastic Agent*, please update the following secrets inside main kustomization.yaml:

Reviewer 2: ```suggestion
For *Managed Elastic Agent*, please update the following secrets inside main [kustomization.yaml](./elastic-agent-managed/kustomization.yaml):
```

- ${fleet_url}: Fleet Server URL to enroll the Elastic Agent into. FLEET_URL can be found in Kibana, go to Management > Fleet > Settings
- ${enrollment_token}: Elasticsearch API key used to [enroll Elastic Agents](https://www.elastic.co/guide/en/fleet/current/fleet-enrollment-tokens.html#fleet-enrollment-tokens) in Fleet.

```yaml
secretGenerator:
    - name: elastic-agent-creds
      literals:
        - host=${fleet_url}
        - api_key=${enrollment_token}
```

> For *Standalone Elastic Agent*, please update the following secrets inside main [kustomization.yaml](./elastic-agent-managed/kustomization.yaml):

Reviewer 2: ```suggestion
For *Standalone Elastic Agent*, please update the following secrets inside main [kustomization.yaml](./elastic-agent-standalone/kustomization.yaml):
```

- ${es_host}: The Elasticsearch host to communicate with
- ${api_key}: The API Key with access privilleges to connect to Elasticsearch. See [create-api-key-standalone-agent](https://www.elastic.co/guide/en/fleet/current/grant-access-to-elasticsearch.html#create-api-key-standalone-agent)

## Remote usage of kustomize templates

Users can use following commands:

Managed Elastic Agent:

> ```bash

Reviewer 2: Would this be the command our users would get from this Kibana page?

![image](https://github.com/elastic/elastic-agent/assets/113898685/b9e8ad57-7a10-4660-b6fc-8c67b6edd833)

Reviewer 2: I think a more intuitive approach would be to change the `kustomization.yaml` file directly. So the user would see this file in Kibana:

```yaml
apiVersion: kustomize.config.k8s.io/v1beta1
kind: Kustomization

namespace: kube-system

images:
  - name: docker.elastic.co/beats/elastic-agent
    newTag: ${IMAGE_TAG}

secretGenerator:
    - name: elastic-agent-creds
      literals:
        - host=${ES_HOST} 
        - enrollment_token=${ENROLLMENT_TOKEN} 

resources:
  - ./base
  - https://github.com/kubernetes/kube-state-metrics

patches:
- path: environmental-variables-remove.yaml
- target:
    group: apps
    version: v1
    kind: DaemonSet
    name: elastic-agent
  path: fleet-enrollment-token-patch.yaml
```

And all `${...}` variables should have literal values by that point.

I think the problem with that command is that it still requires us to evaluate `base64` of the variables and we end up not making use of the literal of the secret map. And then the `sed` is not easy to read.

Author: All of this will be replaced by onliner command and nothing will be visible to the user
